### PR TITLE
ENH:signal: Speed-up Cython implementation of _sosfilt

### PR DESCRIPTION
**Pull request by ilayn**
As per the discussion on #12970, `sosfilt` implementation is accelerated through some rearrangements of the loops. Before this change as shown in the issue the speed difference was occasionally hitting x2 slowdowns.

Continuing the example from the linked issue for `dt=[10, 100, 1000]`  I get the following "almost" equal performance, measured through 

```python
%timeit timeseries_lp_sos = scipy.signal.sosfilt(lp_sos, timeseries)
%timeit timeseries_lp = scipy.signal.lfilter(lp_numerator, lp_denominator, timeseries)
```

```bash
dt = 10
6.74 ms ± 108 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) # This PR
6.1 ms ± 75.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each) # ba performance

dt = 100
67.1 ms ± 940 µs per loop (mean ± std. dev. of 7 runs, 10 loops each) # This PR
63.6 ms ± 932 µs per loop (mean ± std. dev. of 7 runs, 10 loops each) # ba performance

dt = 1000
640 ms ± 9.23 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # This PR
595 ms ± 9.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each) # ba performance
```

Pinging @RonaldAJ ; could I please bother you for a verification of these results?

### CODE REVIEW

**Comment by larsoner** [comment]
On `master`:
```
In [8]: b, a = scipy.signal.iirdesign(0.2,0.3,2,80)
In [9]: sos = scipy.signal.iirdesign(0.2,0.3,2,80, output='sos')
In [10]: x = np.random.RandomState(0).randn(44100 * 100)                                                                                                                            
In [11]: %timeit scipy.signal.lfilter(b, a, x)                                                                                                                                      
30.8 ms ± 639 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [12]: %timeit scipy.signal.sosfilt(sos, x)                                                                                                                                       
38.2 ms ± 236 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
And on this PR:
```
In [15]: %timeit scipy.signal.lfilter(b, a, x)                                                                                                                                      
31.1 ms ± 250 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
In [16]: %timeit scipy.signal.sosfilt(sos, x)                                                                                                                                       
32 ms ± 163 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
CIs are happy and I think we can trust existing tests. Thanks @ilayn !

File: scipy/signal/_sosfilt.pyx [hunk 1]
```diff
@@ -21,38 +24,63 @@ ctypedef fused DTYPE_t:
...
+                # Use direct II transposed structure:
+                x[i, n] = sos[s, 0] * x_n + zi[i, s, 0]
+                zi[i, s, 0] = (sos[s, 1] * x_n - sos[s, 4] * x[i, n] +
+                               zi[i, s, 1])
+                zi[i, s, 1] = (sos[s, 2] * x_n - sos[s, 5] * x[i, n])
+
+
+cpdef void _sosfilt(DTYPE_t [:, ::1] sos,
+             DTYPE_t [:, ::1] x,
+             DTYPE_t [:, :, ::1] zi) nogil:
```
Comment (peterbell10) [inline scipy/signal/_sosfilt.pyx:81]: This doesn't release the GIL any more. Quoting the [docs](https://cython.readthedocs.io/en/latest/src/userguide/external_C_code.html#declaring-a-function-as-callable-without-the-gil)
>The `nogil` function annotation declares that it is safe to call the function without the GIL. It is perfectly allowed to execute it while holding the GIL. The function does not in itself release the GIL if it is held by the caller.

I'll send a follow-up PR.

Comment (larsoner) [reply scipy/signal/_sosfilt.pyx:81]: Thanks @peterbell10 , I missed that!

**Comment by RonaldAJ** [comment]
@ilayn I need to find a suitable moment, if it is still needed. With what I can see in the chat now I am happy with the changes!

**Comment by RonaldAJ** [comment]
Using the script from issue #12970 with an extra line for printing the scipy version I have:

scipy.__version__: 1.6.0.dev0+06f51e3
time elapsed  for ba-filter: 0.06401501893997193
time elapsed for sos-filter: 0.056847667694091795

So indeed performance is now comparable for these settings.

I found however that performance for shorter signals is considerably better (as in shorter computation time per sample). I will dive into it a bit more and make a new issue for it.

**Comment by RonaldAJ** [comment]
For true comparison with the previous scipy version built on the same system:
scipy.__version__: 1.6.0.dev0+9d60a26
time elapsed  for ba-filter: 0.04203344583511352
time elapsed for sos-filter: 0.05560921430587769

**Comment by ilayn** [comment]
@RonaldAJ Thanks. Good to hear that it is confirmed.

**Pull request merged**